modules/host/network.go

```diff
@@ -174,6 +161,20 @@ func (h *Host) threadedListen() {
 	}
 	defer h.tg.Done()
 
+	// Automatically close the listener when h.tg.Stop() is called.
+	err = h.tg.Add()
+	if err != nil {
+		return
+	}
+	go func() {
+		defer h.tg.Done()
+		<-h.tg.StopChan()
+		err = h.listener.Close()
+		if err != nil {
+			h.log.Println("Closing the listener failed:", err)
+		}
+	}()
+
```

Comment:
This thread is not the thread that creates the listener, but this piece of code is the code that creates the listener, therefore this code definitely doesn't make sense here. It also brings back the non-clean shutdown bug that was fixed in the first commit.